lib/ansible/plugins/vars/sops_vars.py

```diff
@@ -0,0 +1,258 @@
+    51: "SopsMacMismatch",
+    52: "SopsMacNotFound",
+    61: "SopsConfigFileNotFound",
+    85: "SopsKeyboardInterrupt",
+    91: "SopsInvalidTreePathFormat",
+    100: "SopsNoFileSpecified",
+    128: "SopsCouldNotRetrieveKey",
+    111: "SopsNoEncryptionKeyFound",
+    200: "SopsFileHasNotBeenModified",
+    201: "SopsNoEditorFound",
+    202: "SopsFailedToCompareVersions",
+    203: "SopsFileAlreadyEncrypted"
+}
+
+
+class SopsError(AnsibleError):
+    ''' extend AnsibleError class with sops specific informations '''
+
+    def __init__(self, filename, exit_code, message="Unknown error",):
+        message = "error with file %s: sops exited with code %d: %s" % (filename, exit_code, message)
+        super(SopsError, self).__init__(message=message)
+
+
+class SopsErrorGeneric(SopsError):
+    pass
+
+
+class SopsCouldNotReadInputFile(SopsError):
+    pass
+
+
+class SopsCouldNotWriteOutputFile(SopsError):
+    pass
+
+
+class SopsErrorDumpingTree(SopsError):
+    pass
+
+
+class SopsErrorReadingConfig(SopsError):
+    pass
+
+
+class SopsErrorInvalidKMSEncryptionContextFormat(SopsError):
+    pass
+
+
+class SopsErrorInvalidSetFormat(SopsError):
+    pass
+
+
+class SopsErrorConflictingParameters(SopsError):
+    pass
+
+
+class SopsErrorEncryptingMac(SopsError):
+    pass
+
+
+class SopsErrorEncryptingTree(SopsError):
+    pass
+
+
+class SopsErrorDecryptingMac(SopsError):
+    pass
+
+
+class SopsErrorDecryptingTree(SopsError):
+    pass
+
+
+class SopsCannotChangeKeysFromNonExistentFile(SopsError):
+    pass
+
+
+class SopsMacMismatch(SopsError):
+    pass
+
+
+class SopsMacNotFound(SopsError):
+    pass
+
+
+class SopsConfigFileNotFound(SopsError):
+    pass
+
+
+class SopsKeyboardInterrupt(SopsError):
+    pass
+
+
+class SopsInvalidTreePathFormat(SopsError):
+    pass
+
+
+class SopsNoFileSpecified(SopsError):
+    pass
+
+
+class SopsCouldNotRetrieveKey(SopsError):
+    pass
+
+
+class SopsNoEncryptionKeyFound(SopsError):
+    pass
+
+
+class SopsFileHasNotBeenModified(SopsError):
+    pass
+
+
+class SopsNoEditorFound(SopsError):
+    pass
+
+
+class SopsFailedToCompareVersions(SopsError):
+    pass
+
+
+class SopsFileAlreadyEncrypte(SopsError):
+    pass
+
+
+def decrypt_with_sops(filename):
+    display.vvvv(u"sops --decrypt %s" % filename)
+
+    # Run sops directly as python module is deprecated
+    process = Popen(["sops", "--decrypt", filename], stdout=PIPE, stderr=PIPE)
+    (output, err) = process.communicate()
+    exit_code = process.wait()
+
+    # DO NOT display output
+    # is the decrypted secret and would easily end in logs :)
+    # if output:
+    #     display.vvvv(output)
+
+    # sops logs always to stderr ( stdout is used for file content )
+    if err:
+        display.vvvv(err)
+
+    if exit_code > 0:
+        if exit_code in sops_error_codes.keys():
+            exception_name = sops_error_codes[exit_code]
+            raise globals()[exception_name](filename, exit_code, err)
```

Comment:
I don't think `globals()` should be used this way.

How about allowing `SopsError` to also accept a symbolic error message (i.e. a string), and passing `exception_name` as this string? Then you have a lot of less code, the code is cleaner, and the error list is easier to extend.

Reply:
I agree, I'm a bit confused on the proposed implementation.

Something like this would be what are you suggesting?

```suggestion
            raise SopsError(exception_name, filename, exit_code, err)
```

`SopsError` implementation:

```python
class SopsError(AnsibleError):
    ''' extend AnsibleError class with sops specific informations '''

    def __init__(self, exception_name, filename, exit_code, message="Unknown error",):
        message = "error with file %s: %s exited with code %d: %s" % (filename, exception_name, exit_code, message)
        super(SopsError, self).__init__(message=message)
```

This would make perfect sense and is surely more easier to maintain in the long run!

Reply:
I thought of something slightly different, but this is also much better. It also reduces the number of exceptions to a way shorter list.

Reply:
I provided a slightly different solution, that do not rely on `exception_name` but just on `exit_code`, as it was already there.

Changes in commit 9e868efb4c